src/instamatic/gui/videostream_processor.py

```diff
@@ -0,0 +1,177 @@
+from __future__ import annotations
+
+import io
+from collections import deque
+from contextlib import contextmanager
+from functools import wraps
+from typing import Any, Iterator, Literal, NamedTuple, Optional, Protocol, Union
+
+import numpy as np
+import PIL.Image
+from matplotlib.figure import Figure
+from PIL import Image, ImageDraw
+
+from instamatic.camera.videostream import VideoStream
+
+
+class VideoStreamFrameProtocol(Protocol):
+    """Mimics the `VideoStreamFrame` interface to avoid circular import."""
+
+    auto_contrast: bool = True
+    brightness: float = 1.0
+    display_range: int = 255
+    stream: VideoStream
+
+
+class DeferredImageDraw:
+    """Defer `ImageDraw` method calls: put them in deque, draw using `on`."""
+
+    class Instruction(NamedTuple):
+        """Stores info about `ImageDraw` calls deferred by `__getattr__`."""
+
+        attr_name: str
+        args: tuple
+        kwargs: dict[str, Any]
+
+    def __init__(self, draw: Optional[ImageDraw] = None) -> None:
+        self._drawing = draw if draw else ImageDraw.Draw(Image.new('RGB', (1, 1)))
+        self.instructions: deque[DeferredImageDraw.Instruction] = deque()
+
+    def __getattr__(self, attr_name: str) -> Union[Any]:
+        """Get the first of `self.attr_name` and `self._drawing.attr_name`.
+
+        If the attribute is a _drawing callable, wrap it so that at call
+        it is added to `self.instructions` instead of being executed directly.
+        `DeferredImageDraw.Instruction` instance created this way is returned
+        and can be deleted by calling `self.instructions.remove(instruction)`.
+        """
+        try:
+            attr = object.__getattribute__(self, attr_name)
+        except AttributeError as e:
+            reraise_on_fail = e
+            try:
+                attr = getattr(self._drawing, attr_name)
+            except AttributeError:
+                raise reraise_on_fail
+
+            if callable(attr):
+
+                @wraps(attr)
+                def wrapped(*args, **kwargs) -> DeferredImageDraw.Instruction:
+                    element = self.Instruction(attr_name, args, kwargs)
+                    self.instructions.append(element)
+                    return element
+
+                return wrapped  # do not call attr - delay it until _redraw()
+        return attr  # non-callable attr of self (if exists) or self._drawing
+
+    def on(self, image: Image.Image) -> Image.Image:
+        """Core method: draws all deferred `self.instructions` on image."""
+        self._drawing = ImageDraw.Draw(image)
+        for attr_name, args, kwargs in self.instructions:
+            getattr(self._drawing, attr_name)(*args, **kwargs)
+        return image
+
+    def circle(
+        self,
+        xy: tuple[int, int],
+        radius: float,
+        fill: Union[str, tuple] = None,
+        outline: Union[str, tuple] = None,
+        width: int = 1,
+    ) -> DeferredImageDraw.Instruction:
```

Comment:
Forcing keyword args here for safety.

```suggestion
    def circle(
        self,
        *,
        xy: tuple[int, int],
        radius: float,
        fill: Union[str, tuple] = None,
        outline: Union[str, tuple] = None,
        width: int = 1,
    ) -> DeferredImageDraw.Instruction:
```

Reply:
Initially committed, but then reverted: unfortunately the PIL signature allows all of these parameters to be positional as well as keyword, and the whole point of this method is backwards compatibility with PIL, so `*,` is a no-no.